New metadata API: add MetaFile and TargetFile classes

MVrachev opened the pull request:
This pr is a continuation from pr https://github.com/theupdateframework/tuf/pull/1223

This is the final pr planned for https://github.com/theupdateframework/tuf/issues/1139, 

**Description of the changes being introduced by the pull request**:

In the top-level metadata classes, there are complex attributes such as
"meta" in Targets and Snapshot, "key" and "roles" in Root etc.
We want to represent those complex attributes with a class to allow
easier verification and support for metadata with unrecognized fields.
For more context read ADR 0004 and ADR 0008 in the docs/adr folder.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

Commit:
New API: Add MetaFile class

In the top-level metadata classes, there are complex attributes such as
"meta" in Targets and Snapshot, "key" and "roles" in Root etc.
We want to represent those complex attributes with a class to allow
easier verification and support for metadata with unrecognized fields.
For more context read ADR 0004 and ADR 0008 in the docs/adr folder.

Additionally, after adding the MetaFile class, when we create an object
we are now calling from dict twice - one for the main class (Timestamp,
Snapshot) and one for the pacticular complex attribute -
MetaFile.from_dict(). Given that the "from_dict" methods have the
side effect of destroying the given dictionary, we would need to
start using deepcopy() for our tests.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
New API: Add TargetFile class

In the top-level metadata classes, there are complex attributes such as
"meta" in Targets and Snapshot, "key" and "roles" in Root etc.
We want to represent those complex attributes with a class to allow
easier verification and support for metadata with unrecognized fields.
For more context read ADR 0004 and ADR 0008 in the docs/adr folder.

As written in the spec "targets" in "targets.json" has defined the
"custom" field serving the same purpose as "unrecognized_fields" in the
implementation.
That's why to conform against the spec and support "custom" and allow
"unrecognized_fields" everywhere where it's not sensitive we can define
custom as property which actually access data stored in
unrecognized_fields.
For context read ADR 8 in tuf/docs/adr.

Additionally, after adding the TargetFile class, when we create a
Targets an object we are now calling from dict twice - one for the main
Targets class and one for each of the complex attributes
TargetFile.from_dict() and Delegations.from_dict().
Given that the "from_dict" methods have the side effect of destroying
the given dictionary, we would need to start using deepcopy()
for our tests.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

sechkova commented on tuf/api/metadata.py:
```diff
@@ -463,22 +463,64 @@ def remove_key(self, role: str, keyid: str) -> None:
+class MetadataInfo:
+    """A container with information about a particular metadata file.
+    Instances of MetadataInfo are used as values in a dictionary called
+    "meta" in Timestamp and Snapshot.
```
Can you add a new line before the 'Attributes' section?

sechkova commented on tuf/api/metadata.py:
```diff
@@ -463,22 +463,64 @@ def remove_key(self, role: str, keyid: str) -> None:
+
+    def to_dict(self) -> Dict:
+        """Returns the dictionary representation of self. """
+        json_dict = {"version": self.version}
```
After abstracting the (de)serialization, other formats than `json` are possible and although the default is `json` it shouldn't be hardcoded.

sechkova commented on tuf/api/metadata.py:
```diff
@@ -463,22 +463,64 @@ def remove_key(self, role: str, keyid: str) -> None:
-                        ...
-                    } // optional
-                }
+            'meta': {
```
'meta' is already listed as class attribute above.

sechkova commented on tuf/api/metadata.py:
```diff
@@ -489,38 +531,39 @@ def __init__(
         common_args = cls._common_fields_from_dict(timestamp_dict)
-        meta = timestamp_dict.pop("meta")
+        meta = {
+            "snapshot.json": MetadataInfo(
```
Same comment as above, related to hardcoding file formats. In this case if an implementation does not use `.json` files this will probably raise a `KeyError`.
Possibly the filename/format should be read from `timestamp_dict` instead.

MVrachev replied:
That's a really good catch!
Will make it work with other extensions as well!

sechkova commented on tuf/api/metadata.py:
```diff
@@ -628,13 +622,7 @@ def update(
     ) -> None:
         """Assigns passed (delegated) targets role info to meta dict. """
         metadata_fn = f"{rolename}.json"
```
This is not your change but again, hardcoded `.json`.

sechkova commented on tuf/api/metadata.py:
```diff
@@ -463,22 +463,64 @@ def remove_key(self, role: str, keyid: str) -> None:
+        self.length = length
+        self.hashes = hashes
+
+    def __eq__(self, other: "MetadataInfo") -> bool:
```
The `__eq__` implementation is needed only for the tests, right?
I guess it doesn't hurt having it :shrug: Maybe it will come handy in future.

MVrachev replied:
Yes, currently `__eq__` is needed only for tests.

sechkova commented on tuf/api/metadata.py:
```diff
@@ -625,22 +625,64 @@ def update(
+
+    def to_dict(self) -> Dict:
+        """Returns the JSON-serializable dictionary representation of self. """
+        json_dict = {"length": self.length, "hashes": self.hashes}
```
Same comment about `json`.

sechkova reviewed (commented):
In the PR description you say that it will:
> Rename a few instances of the update method in our classes and use a more verbose name.

but there is only one commit with marked TODOs.

My general comment is about removing hardcoded file extensions and mentions of "json" throughout the code. After the abstract (de)serialization was implemented, TUF supposedly supports other file formats too.

MVrachev commented:
I rebased and fixed your two small comments @sechkova about other issues besides the hard-coded file names.

I fixed the problem of hardcoded `json` strings partially in the places I modified already.
To fully fix this problem we either have to use a more complicate solution or wait and consider other alternatives
which will give us the ability to set metadata extensions as a configuration option.

Created an issue about the hardcoded filename problems: https://github.com/theupdateframework/tuf/issues/1333

sechkova commented on tuf/api/metadata.py:
```diff
@@ -534,22 +534,41 @@ def __init__(
         self.meta = meta
 
+    @classmethod
+    def _get_snapshot_filename(cls, meta_dict):
```
This is a reasonable check but aren't we just assuming that input data is correct everywhere else? Since input validation is still a WIP? Anyway, probably better having it than not.

MVrachev replied:
Well the problem is that `meta_dict` sometimes is passed from the user - when calling `Timestamp.from_dict()`
and consequently when calling the `cls. _get_snapshot_filename()` method we haven't validated the `meta_dict` field.

At least, that is the case before implementing any new validation option as described in ADR 007.

jku commented on tuf/api/metadata.py:
```diff
@@ -458,22 +460,65 @@ def remove_key(self, role: str, keyid: str) -> None:
+        self,
+        version: int,
+        length: Optional[int] = None,
+        hashes: Optional[Dict] = None,
```
`Optional[Dict[str, str]]` I suppose

(much of the API uses Mapping in place of Dict though... I think that might be a mistake but maybe I've missed the reasons?)

jku commented on tuf/api/metadata.py:
```diff
@@ -484,35 +529,61 @@ def __init__(
+            snapshot_filename: MetadataInfo(
+                **timestamp_dict["meta"][snapshot_filename]
+            )
+        }
```
I think `get_snapshot_filename()` is not useful -- it does not do much and the ValueError is now hidden in an unexpected function. Overall we end up doing many (4?) dictionary lookups when I think one would be enough (it's not a performance issue or that important here but ... just more code than should be needed).

jku replied:
Something like this (assuming the attribute changes I proposed above, and popping being acceptable):
```
meta_dict = timestamp_dict["meta"]
# TODO check length
snapshot_filename, snapshot_dict = meta_dict.popitem()
snapshot_info = MetadataInfo(**snapshot_dict)
return cls(*common_args, snapshot_filename, snapshot_info)
```

MVrachev replied:
I removed `get_snapshot_filename` from the latest rebased version.
I will think more about how to fix this approach when working on #1333

jku commented on tuf/api/metadata.py:
```diff
@@ -484,35 +529,61 @@ def __init__(
         spec_version: str,
         expires: datetime,
-        meta: Mapping[str, Any],
+        meta: Mapping[str, MetadataInfo],
```
We could just have `snapshot_filename` and `snapshot_info` attributes instead of the dictionary... 

That way the API would be more usable as well: `timestamp.signed.meta["snapshot.json"].version` is ugly and won't autocomplete whereas `timestamp.signed.snapshot_info.version` is ... well, harder to typo at least.

jku replied:
the above comment was done when I wasn't aware/thinking of the earlier decision to try to map the API to file format as close as possible... in that light you might want to disregard this (I think I still want the usable version but that can be added later as a property)

MVrachev replied:
Agree, this could be subject to another set of changes.

jku commented on tuf/api/metadata.py:
```diff
@@ -548,23 +606,29 @@ def __init__(
         common_args = cls._common_fields_from_dict(snapshot_dict)
-        meta = snapshot_dict.pop("meta")
+        meta = {}
+        for meta_path in snapshot_dict["meta"].keys():
```
`for meta_path, meta_dict in snapshot_dict["meta"].items()` -- no need to do the lookup later

jku commented on tuf/api/metadata.py:
```diff
@@ -484,35 +529,61 @@ def __init__(
+        version: int,
+        length: Optional[int] = None,
+        hashes: Optional[Mapping[str, Any]] = None,
     ) -> None:
```
I don't know where this is used so this may be dumb, but is this function needed?
Would `timestamp.signed.snapshot_info = MetadataInfo(version, length, hashes)` be equivalent to to calling this ?

(this assumes the snapshot_info change proposed above)

MVrachev replied:
It will be the same I suppose, but I hope we will use this function to make some validation before doing the assignment.

jku replied:
I think a property checks all boxes: I'll leave a comment in the new code

jku commented on tuf/api/metadata.py:
```diff
@@ -650,28 +752,38 @@ def __init__(
-        targets = targets_dict.pop("targets")
-        delegations = targets_dict.pop("delegations")
+        targets = {}
+        for target_path in targets_dict["targets"].keys():
```
`for target_path, target_dict in targets_dict["targets"].items():`

jku commented on tuf/api/metadata.py:
```diff
@@ -680,4 +792,8 @@ def to_dict(self) -> Dict[str, Any]:
     # Modification.
     def update(self, filename: str, fileinfo: Mapping[str, Any]) -> None:
```
Same question here, is this a useful function vs caller modifying targets.signed.targets directly

MVrachev replied:
It will be the same I suppose, but I hope we will use this function to make some validation before doing the assignment.

jku replied:
good point, thanks

jku commented on tuf/api/metadata.py:
```diff
@@ -484,35 +529,61 @@ def __init__(
         timestamp_dict = self._common_fields_to_dict()
-        timestamp_dict.update({"meta": self.meta})
+        snapshot_filename = self._get_snapshot_filename(self.meta)
+        timestamp_dict.update(
+            {
+                "meta": {
+                    snapshot_filename: self.meta[snapshot_filename].to_dict()
+                }
+            }
+        )
```
This works but is there a reason to use update over the simpler index access when you only have a single item?
```
 timestamp_dict["meta"] = {
     ...
}
```

jku reviewed (commented):
Apologies for not looking at this before. I've gone through most, but maybe not quite processed all of it.

Martin knows of this but for others information: the changes to `.pop()` usage are going to clash with another PR #1345: they need to be coordinated. In general I think this PR could have been several smaller instead -- I get that it means annoying depending PRs... but this way means I feel bad for leaving so many comments :grin:  

Would be great to see a bit more prose about the design and usability of the specific API classes somewhere: If we go through the trouble of creating objects for everything, we should try to make them more usable than the dictionaries they replace. As examples, see question in code about snapshot_info attribute and ideas like _'should Targets maybe implement abc.Mapping so caller could access targets.signed["target_name"] instead of targets.signed.targets["target_name"]'_ -- that last one may not be a good idea but it would be good to see some ideas and discussion before we start maintaining more code.

This feels like a lot to ask on a big PR but would be nice to see how the code now copes without broken input in tests (missing fields, extra fields, etc) -- to confirm the benefits that we should be getting from this work.

jku commented:
> Apologies for not looking at this before. I've gone through most, but maybe not quite processed all of it.

Also grumpy yesterday me sounds like he is unsatisfied with this PR but that's not true so I'll say it now:
PR looks correct to me (no bugs that I can find), good work, thank you

jku commented:
> Would be great to see a bit more prose about the design and usability of the specific API classes somewhere: If we go through the trouble of creating objects for everything, we should try to make them more usable than the dictionaries they replace

After further discussion, this opinion seems to go against previous decisions to keep the API as close to file format as possible -- so possibly we need to discuss further or implement this as "faithful reproduction of spec" (disregarding my comments) and think about potential usability improvements afterwards...

MVrachev commented:
There were a lot of discussions after my last update on this pr.
I would prefer if we wait for #1345 and #1360 to be merged first.
Then, I will update/split this pr.

Commit:
Add MetaFile/TargetFile specific tests

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Disable "C0302" for tuf/api/metadata.py

Disable the "C0302: Too many lines in module" warning which warns for modules
with more 1000 lines, because all of the code here is logically connected
and currently, we are above 1000 lines by a small margin.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I have rebased and made changes in this pr after "develop" has evolved a lot.

Some of the changes I made:
- rename `MetadataInfo` to `MetaFile` and `TargetInfo` to `TargetFile`. Those names seem better in my opinion because they communicate better what the classes represent.
- added support for unrecognized fields in `MetaFile` and `TargetFile`. I didn't add a test, because #1382 will add an automatic test for that for dictionaries and classes.
- dropped a couple of the planned commits after #1367 was merged.
- dropped a couple more commits, because I realized they were not needed - removing the side effect of destroying the dictionaries when calling `from_dict()` from the classes (this is needed for ADR 8) and adding a todo comment about the `update()` functions naming (there is an issue for that #1230)
- added `from_dict()` for `MetaFile` and `TargetFile`
- added `MetaFile` and `TargetFile` specific tests
- disabled the `C0302: Too many lines in module` warning for tuf/api/metadata.py.

@jku and @sechkova pr is again ready for a fresh review.

jku commented on tuf/api/metadata.py:
```diff
@@ -698,25 +741,31 @@ def __init__(
+        meta_dict = snapshot_dict.pop("meta")
+        meta = {}
+        for meta_path in meta_dict.keys():
+            meta[meta_path] = MetaFile.from_dict(meta_dict[meta_path])
```
`for meta_path, metafile_dict in meta_dict.items():` or something like it maybe?

jku commented on tuf/api/metadata.py:
```diff
@@ -698,25 +741,31 @@ def __init__(
+        for meta_path, meta_info in self.meta.items():
+            meta_dict[meta_path] = meta_info.to_dict()
+
+        snapshot_dict.update({"meta": meta_dict})
```
nit: `snapshot_dict["meta"] = meta_dict` seems more normal

jku commented on tuf/api/metadata.py:
```diff
@@ -897,35 +1002,42 @@ def __init__(
+        for target_path, target_file_obj in self.targets.items():
+            targets[target_path] = target_file_obj.to_dict()
+
+        res_dict.update({"targets": targets})
```
`res_dict["targets"] = targets`

jku commented on tuf/api/metadata.py:
```diff
@@ -897,35 +1002,42 @@ def __init__(
     def update(self, filename: str, fileinfo: Dict[str, Any]) -> None:
         """Assigns passed target file info to meta dict."""
-        self.targets[filename] = fileinfo
+        self.targets[filename] = TargetFile.from_dict(fileinfo)
```
This function should have TargetFile as argument, no?

MVrachev replied:
I still think this is not in the scope of this pr.
See my comment https://github.com/theupdateframework/tuf/pull/1329#discussion_r631716043.

jku replied:
I'm fine with not completely re-designing update() methods in this PR.

I disagree with leaving the argument as a dict. Why are we going through the trouble of adding objects and then not using them in the API? 

I think your test code makes the case well: it does:
`targets.signed.update(filename, fileinfo.to_dict())`
and then in update you do 
` TargetFile.from_dict(fileinfo)`
it seems the argument should be TargetFile here

MVrachev replied:
I commented about the same question here:
https://github.com/theupdateframework/tuf/pull/1329#discussion_r632555437

jku commented on tuf/api/metadata.py:
```diff
@@ -861,22 +904,84 @@ def to_dict(self) -> Dict[str, Any]:
+        return (
+            self.length == other.length
+            and self.hashes == other.hashes
+            and self.custom == other.custom
```
missing unrecognised_fields

jku replied:
I see `__eq__()` is still here (which is fine -- maybe you use it in tests) but it's still not complete: it should compare unrecognized_fields instead of custom

jku commented on tuf/api/metadata.py:
```diff
@@ -861,22 +904,84 @@ def to_dict(self) -> Dict[str, Any]:
+        self,
+        length: int,
+        hashes: Dict[str, str],
+        custom: Optional[Dict[str, Any]] = None,
```
It could be argued we don't need to handle custom at all in any of these functions: it's just an unrecognised field and already supported via that mechanism.

If we want to match the file format exactly we'd then need a "custom" property getter though (that refers to `unrecognised_fields.get("custom")`)

No strong opinions, maybe I lean a bit towards the property :shrug:

MVrachev replied:
I like the idea of using custom as `unrecognized fields` through using a property. 
Will do it and will add a comment that custom will replace `unrecognized fields` for this particular case.

MVrachev replied:
Wondering, if we use `custom` replacing the need of `unrecognized fields` for this class, then do you test for this in #1382?
Do you add an unrecognized field under the `custom` dictionary?

jku replied:
I was trying to suggest dropping "custom", and just letting the already existing unrecognised_fields functionality handle it -- that basically handles de/serializing both "custom" and any other unrecognised fields.

Then I was saying that we might still want to add a property "custom" (to conform with the API design principle that the object API would copy the file format) but the property would just access `self.unrecognised_fields.get("custom")` and de/ser still would not handle "custom" at all -- it's handled like other unrecognised fields.

What you implemented is close but not quite the same... now TargetFile cannot have unrecognised fields at all (it can only have a "custom" field): as you correctly point out that means deciding that TargetFile is not allowed to have unrecognised fields and updating the test... which seems like the implementation detail guiding how we read the specification. 

I don't think this is a major issue, but I also think what I suggested would cover all the cases without any more complexity... opinions?

MVrachev replied:
Well, `custom` basically is the same as `unrecognised_fields`.
I feel like we should encourage our users to use `custom` as described in the spec when adding additional fields instead of relying on `unrecognized_fields` at least in this scenario.
Maybe we can achieve that by adding a `custom` property, but then if we want to treat it as a regular attribute maybe we would have to add a property setter as well?

On another hand, if we don't have `unrecognized_fields` in TargetFile, target.json file with unrecognized fields outside custom won't work...

jku replied:
> Well, `custom` basically is the same as `unrecognised_fields`.
> I feel like we should encourage our users to use `custom` as described in the spec when adding additional fields instead of relying on `unrecognized_fields` at least in this scenario.
> Maybe we can achieve that by adding a `custom` property, but then if we want to treat it as a regular attribute maybe we would have to add a property setter as well?
> 
> On another hand, if we don't have `unrecognized_fields` in TargetFile, target.json file with unrecognized fields outside custom won't work...

that's what I'm saying: if we read the spec as "you need to be able to add new fields to any dictionary where it's technically possible" (which was the consensus I believe), then  implementing custom is not enough for spec-compliance

I don't think it's a major thing but ... if we can easily achieve spec compliance in a trivial matter then why not do it?

MVrachev replied:
Okay, will update the class, so that `unrecognized fields` is documented and added as an attribute, but I am not sure should we document `custom` when `custom` will be just a property and we retrieve with `self.unrecognised_fields.get("custom")`?

MVrachev replied:
I didn't define it in my last update. Do you think that's okay?

jku commented on tuf/api/metadata.py:
```diff
@@ -655,13 +717,7 @@ def update(
-
-        if hashes is not None:
-            self.meta["snapshot.json"]["hashes"] = hashes
+        self.meta["snapshot.json"] = MetaFile(version, length, hashes)
```
If we have this function it should take a Metafile as argument, IMO.

some other opinions on update():
* if we keep it a function, it should be renamed: "update" tells me nothing
* on the other hand I kind of think this should be a property instead, maybe "snapshot_meta" so you could do things like :
`version = timestamp.signed.snapshot_meta.version
`
or 
`timestamp.signed.snapshot_meta = MetaFile(ver, len, hashes)`

jku replied:
there may be an argument _against_ using helper properties like the one I handwaved: that it does not separate the attributes-that-map-the-file-format and the helpers enough (and that functions like `update()` are better in that regard). 

I think that's an issue we should consider but it does not outweigh the advantages (such as having a single read-and-write property with a descriptive name).

MVrachev replied:
About the `update()` naming there is an issue already opened capturing this #1230.

I think we should restrain ourselves from changing the API here (like changing the argument to `MetaFile` or adding properties), but focus more on the changes included in the pr. 

Instead, those could be discussed in the issue I referenced above and the issue area of focus could widen to discuss the `update()` functions design as a whole.

jku replied:
I'm fine with not re-designing update here.

I disagree with leaving the argument as a dict. Why are we going through the trouble of adding objects and then not using them in the API?

jku replied:
Leaving update() re-design for later is fine.

I see no reason (so far) not to make the argument a MetaFile.

MVrachev replied:
Okay, will change the argument type for the `update` functions, but won't rename it or add additional attributes as you suggested above.

jku commented:
A question on the `__eq__` implementations: do we need them? None of the other classes seem to have them -- I assumed you used them in the tests but that does not seem the case (or maybe I just missed it)

MVrachev commented:
Updated the pr addressing almost all of the comments @jku made with the only exceptions when commenting on the `update` functions.
There I feel the changes Jussi proposed are out of scope for this pr as I described here https://github.com/theupdateframework/tuf/pull/1329#discussion_r631716043.

Additionally, I added a new commit where I remove the `__eq__` methods.

Commit:
Change "update()" argument types

Currently, when we call Targets/Snapshot/Timestamp.update() we are
passing all of the necessary values to create MetaFile/Targets File
respectively.
This is not needed, given that one of the reasons we have created
MetaFile and TargetFile is to make the API easier to use.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

jku commented on tuf/api/metadata.py:
```diff
@@ -624,42 +675,33 @@ def __init__(
         """Returns the dict representation of self."""
         timestamp_dict = self._common_fields_to_dict()
-        timestamp_dict.update({"meta": self.meta})
+        timestamp_dict.update(
+            {"meta": {"snapshot.json": self.meta["snapshot.json"].to_dict()}}
+        )
```
please, no need to use update when a simple `timestamp_dict["meta"] = ...` works

jku commented on tuf/api/metadata.py:
```diff
@@ -859,22 +882,69 @@ def to_dict(self) -> Dict[str, Any]:
+            "hashes": self.hashes,
+        }
+
+        if self.custom is not None:
+            res_dict["custom"] = self.custom
```
Should you have unrecognised_fields handling here instead of handling custom?

MVrachev replied:
I realized that I don't need to handle custom specifically anymore.
"custom", if exist, will be part of `unrecognized fields and my `to_dict` function could be just:
```
    def to_dict(self) -> Dict[str, Any]:
        """Returns the JSON-serializable dictionary representation of self."""
        return {
            "length": self.length,
            "hashes": self.hashes,
            **self.unrecognized_fields,
        }
```

jku commented on tests/test_api.py:
```diff
@@ -241,33 +243,83 @@ def test_metadata_base(self):
         snapshot = Metadata.from_file(snapshot_path)
 
         # Create a dict representing what we expect the updated data to be
-        fileinfo = copy.deepcopy(snapshot.signed.meta)
+        fileinfo = copy.deepcopy(snapshot.signed.meta['role1.json'])
```
This is not a dict like comment says but a MetaFile object, I think?

Is there a reason for this deepcopy and then modifying data -- couldn't you just create the object with MetaFile() using your arguments to make it clear what it is?

jku commented on tests/test_api.py:
```diff
@@ -241,33 +243,83 @@ def test_metadata_base(self):
+        fileinfo = MetaFile(3)
+        snapshot.signed.update('role2', fileinfo)
+        self.assertEqual(
+            snapshot.signed.meta['role2.json'].to_dict(), fileinfo.to_dict()
```
Just a nit but mentioning in case there's something missing: This test doesn't really test much anything anymore (it used to compare two separate dicts, now it's pretty much comparing an object to itself)

MVrachev replied:
Yes, I understand this. The test by itself makes sure that we are actually supporting optional `hashes` and `length`.
Will remove the test in the `fix update argument` commit. Then, we won't need this, because length and hashes will be optional because of the `MetaFile` implementation, no the update function itself.

jku commented on tests/test_api.py:
```diff
@@ -297,26 +349,32 @@ def test_metadata_timestamp(self):
-
-        self.assertNotEqual(timestamp.signed.meta['snapshot.json'], fileinfo)
-        timestamp.signed.update(2, 520, hashes)
-        self.assertEqual(timestamp.signed.meta['snapshot.json'], fileinfo)
+        fileinfo.hashes = hashes
+        fileinfo.version = 2
+        fileinfo.length = 520
```
nit, This would be much more understandable as `fileinfo = MetaInfo(...)` as well

jku commented on tests/test_api.py:
```diff
@@ -297,26 +349,32 @@ def test_metadata_timestamp(self):
-        timestamp.signed.update(3)
-        fileinfo = {'version': 3}
-        self.assertEqual(timestamp.signed.meta['snapshot.json'], fileinfo)
+        fileinfo = MetaFile(version=3)
+        timestamp.signed.update(fileinfo)
+        self.assertEqual(
+            timestamp.signed.meta['snapshot.json'].to_dict(), fileinfo.to_dict()
+        )
```
same nit as before, this doesn't test much anything anymore

jku reviewed (changes requested):
Everything else is a style comment (feel free to disregard) but I think there's a bug in the "custom" handling

Commit:
Remove testing "update()" with "version" only

We have tests which make sure we can use `Timestamp.update()` and
`Snapshot.update()` with MetaFile instance storing only version
(because length and hashes are optional).
Those tests were created to make sure that we are actually supporting
optional hashes and length when we call `update` for those classes, but
after we changed the `update()` signature to accept `MetaFile` instance
the tests are obsolete.
The reason is that length and hashes can be optional because of the
MetaFile implementation, no the update function itself and we have
other tests validating creating a MetaFie instance without hashes and
length.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
Addressed all of @jku commets, and additionally I wanted to make sure that each of the commits passes the CI and makes sense.
That's why I also made the following changes:
1. Merge changes in "Replace usage of dict.copy with copy.deepcopy()" in "Add MetaFile" and "Add TargetFile". Then drop "Replace usage of dict.copy with copy.deepcopy()"
2. Simplified how we handle "custom" in `TargetFile.to_dict()`
3. Squash "New API: represent custom with unrecognized_fields" into "Add TargetFile"
4. Added a new commit removing two tests testing "update with only version" because they are obsolete.

jku reviewed (approved):
I think my comment in-source won't show up since it's in a resolved discussion but:

The` __eq__()` implementation is still not complete: if it is there, it should compare unrecognized_fields instead of custom.

Altogether this looks good, thanks. The docstrings are very verbose but let's trim those later... I'm marking this approved already, but please fix `__eq__()` still in this PR.

MVrachev commented:
> I think my comment in-source won't show up since it's in a resolved discussion but:
> 
> The` __eq__()` implementation is still not complete: if it is there, it should compare unrecognized_fields instead of custom.
> 
> Altogether this looks good, thanks. The docstrings are very verbose but let's trim those later... I'm marking this approved already, but please fix `__eq__()` still in this PR.

I thought I had removed the `__eq__` methods.
It was a mistake that they are left there.
Removed them and rebased.

jku commented:
I was about to merge but now it has conflicts with a bug fix I merged yesterday. Sorry. Could you rebase?

MVrachev commented:
Done.
